File: verify/verify.py
# Recompute the four published AUCs from the per-row scores and verify the
# drift table's internal consistency.
#
# The AUC uses the same Mann-Whitney rank-sum formula as the C, R, and SQL
# verifiers. Printing the canonical AUC| lines lets verify.sh compare all
# implementations to twelve decimals.
#
# The drift check is new: ea_drift.csv reports a per-week AUC and fraud rate,
# and this verifier confirms that every row's fraud rate and AUC are within
# the bounds you would expect from the reported sample sizes (fraud rate is a
# proportion of n, so it must be expressible as k/n for some integer k).
#
# No dependencies beyond the standard library.

import csv
import math
import os
import sys

root = sys.argv[1] if len(sys.argv) > 1 else "."

PUB_TOL = 5e-5


def read_csv(name):
    with open(os.path.join(root, "reports", name), newline="") as f:
        return list(csv.DictReader(f))


# -- AUC via mid-ranks (Mann-Whitney) ----------------------------------------

def auc(ys, scores):
    n = len(ys)
    indexed = sorted(range(n), key=lambda i: scores[i])
    rank = [0.0] * n
    i = 0
    while i < n:
        j = i
        while j + 1 < n and scores[indexed[j + 1]] == scores[indexed[i]]:
            j += 1
        mid = (i + 1 + j + 1) / 2.0
        for k in range(i, j + 1):
            rank[indexed[k]] = mid
        i = j + 1
    n_pos = sum(1 for y in ys if y == 1)
    n_neg = sum(1 for y in ys if y == 0)
    if n_pos == 0 or n_neg == 0:
        return float("nan")
    rank_sum = sum(rank[i] for i in range(n) if ys[i] == 1)
    return (rank_sum - n_pos * (n_pos + 1) / 2) / (n_pos * n_neg)


scores_rows = read_csv("validation_gap_scores_synthetic.csv")
published = read_csv("validation_gap_synthetic.csv")

# Group rows by (split, target encoding, fold)
groups = {}
for row in scores_rows:
    key = (row["split"], row["target encoding"], int(row["fold"]))
    groups.setdefault(key, ([], []))
    groups[key][0].append(int(row["y"]))
    groups[key][1].append(float(row["score"]))

failures = 0

print("point estimates, against reports/validation_gap_synthetic.csv")
for pub in published:
    split = pub["split"]
    te = pub["target encoding"]
    want = float(pub["AUC"])
    fold_aucs = []
    for (s, t, f), (ys, sc) in sorted(groups.items()):
        if s == split and t == te:
            fold_aucs.append(auc(ys, sc))
    got = sum(fold_aucs) / len(fold_aucs)
    delta = abs(got - want)
    ok = delta <= PUB_TOL
    if not ok:
        failures += 1
    status = "ok" if ok else "FAIL"
    print(f"  {split:<16s} {te:<11s} AUC {got:.12f}  published {want:.4f}  |d| {delta:.1e}  {status}")
    print(f"AUC|{split}|{te}|{got:.12f}")

# -- drift table consistency -------------------------------------------------

print("\ndrift table, reports/ea_drift.csv")
drift = read_csv("ea_drift.csv")
drift_checks = 0
for row in drift:
    n = int(row["n"])
    rate = float(row["fraud rate"])
    # The fraud rate should be expressible as round(k/n, 4) for some integer k
    k = round(rate * n)
    reconstructed = round(k / n, 4)
    d = abs(reconstructed - rate)
    ok = d < 1e-4
    drift_checks += 1
    if not ok:
        failures += 1
        print(f"  FAIL week {row['week']}: fraud rate {rate} not consistent with n={n}")
    # AUC should be between 0 and 1
    a = float(row["AUC"])
    drift_checks += 1
    if not (0.0 <= a <= 1.0):
        failures += 1
        print(f"  FAIL week {row['week']}: AUC {a} out of range")

if drift_checks > 0 and failures == 0:
    print(f"  {drift_checks} checks on {len(drift)} weeks, all consistent")

if failures > 0:
    print(f"\n{failures} checks failed")
    sys.exit(1)

print("\nPython reproduces every published AUC and confirms drift-table consistency")

File: verify/verify.rb
# Recompute the four published AUCs from the per-row scores and verify that
# the submission file covers the expected ID range without duplicates.
#
# The AUC uses the same Mann-Whitney mid-rank formula as every other verifier,
# so verify.sh can require all implementations to agree to twelve decimals.
#
# The submission check is new: reports/submission.csv should have one row per
# TransactionID, no duplicates, and every predicted probability in [0, 1].
#
# No gems required.

require "csv"

root = ARGV[0] || "."

PUB_TOL = 5e-5

def read_csv(root, name)
  CSV.read(File.join(root, "reports", name), headers: true)
end

# -- AUC via mid-ranks (Mann-Whitney) ----------------------------------------

def auc(ys, scores)
  n = ys.length
  indexed = (0...n).sort_by { |i| scores[i] }
  rank = Array.new(n, 0.0)
  i = 0
  while i < n
    j = i
    while j + 1 < n && scores[indexed[j + 1]] == scores[indexed[i]]
      j += 1
    end
    mid = (i + 1 + j + 1) / 2.0
    (i..j).each { |k| rank[indexed[k]] = mid }
    i = j + 1
  end
  n_pos = ys.count(1)
  n_neg = ys.count(0)
  return Float::NAN if n_pos == 0 || n_neg == 0
  rank_sum = (0...n).select { |i| ys[i] == 1 }.sum { |i| rank[i] }
  (rank_sum - n_pos * (n_pos + 1) / 2.0) / (n_pos * n_neg)
end

scores_rows = read_csv(root, "validation_gap_scores_synthetic.csv")
published = read_csv(root, "validation_gap_synthetic.csv")

# Group by (split, target encoding, fold)
groups = Hash.new { |h, k| h[k] = [[], []] }
scores_rows.each do |row|
  key = [row["split"], row["target encoding"], row["fold"].to_i]
  groups[key][0] << row["y"].to_i
  groups[key][1] << row["score"].to_f
end

failures = 0

puts "point estimates, against reports/validation_gap_synthetic.csv"
published.each do |pub|
  split = pub["split"]
  te = pub["target encoding"]
  want = pub["AUC"].to_f
  fold_aucs = groups.keys
    .select { |s, t, _f| s == split && t == te }
    .sort
    .map { |key| auc(groups[key][0], groups[key][1]) }
  got = fold_aucs.sum / fold_aucs.length
  delta = (got - want).abs
  ok = delta <= PUB_TOL
  failures += 1 unless ok
  status = ok ? "ok" : "FAIL"
  printf "  %-16s %-11s AUC %.12f  published %.4f  |d| %.1e  %s\n",
         split, te, got, want, delta, status
  printf "AUC|%s|%s|%.12f\n", split, te, got
end

# -- submission file checks ---------------------------------------------------

puts "\nsubmission file, reports/submission.csv"
sub = read_csv(root, "submission.csv")
sub_checks = 0

ids = sub.map { |r| r["TransactionID"].to_i }
sub_checks += 1
dups = ids.length - ids.uniq.length
if dups > 0
  failures += 1
  puts "  FAIL #{dups} duplicate TransactionIDs"
else
  puts "  #{ids.length} rows, no duplicate IDs"
end

bad_probs = 0
sub.each do |row|
  p = row["isFraud"].to_f
  bad_probs += 1 unless p >= 0.0 && p <= 1.0
end
sub_checks += 1
if bad_probs > 0
  failures += 1
  puts "  FAIL #{bad_probs} predictions outside [0, 1]"
else
  puts "  all predictions in [0, 1]"
end

if failures > 0
  puts "\n#{failures} checks failed"
  exit 1
end

puts "\nRuby reproduces every published AUC and confirms submission-file integrity"
